{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[![](https://img.shields.io/badge/Source%20on%20GitHub-orange)](https://github.com/laminlabs/lamin-spatial/blob/main/docs/vitessce.ipynb) [![hub](https://img.shields.io/badge/View%20on%20lamin.ai-mediumseagreen)](https://lamin.ai/vitessce/examples/transform/BZhZQ6uIbkWv)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "nbsphinx": "hidden"
   },
   "source": [
    "# Vitessce: SpatialData"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This tutorial demonstrates how to create interactive Vitessce visualizations for SpatialData objects stored in LaminDB artifacts.\n",
    "\n",
    "We'll work with spatial transcriptomics data in Zarr, OME-TIFF, and OME-Zarr formats."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "!lamin connect laminlabs/lamin-dev  # <-- replace with your instance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "import vitessce as vit\n",
    "import lamindb as ln\n",
    "\n",
    "ln.track()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create the central query object for the public [vitessce/examples](https://lamin.ai/vitessce/examples) instance:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "db = ln.DB(\"vitessce/examples\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "## Visualize a SpatialData object (Zarr format)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "Here we use an example Visium dataset that has been previously ingested into the public [vitessce/examples](https://lamin.ai/vitessce/examples) instance in this [transform](https://lamin.ai/vitessce/examples/transform/y4tFWARW6bJ1)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "sdata_zarr_artifact = db.Artifact.get(key=\"vitessce_examples/visium.sdata.zarr\").save()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "### Save a VitessceConfig object"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "You can create a dashboard for one or several datasets by using Vitessce's component API.\n",
    "Here, we use the [SpatialDataWrapper](https://python-docs.vitessce.io/api_data.html#vitessce.wrappers.SpatialDataWrapper) class to specify which parts of the SpatialData object will be loaded for visualization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "vc = vit.VitessceConfig(\n",
    "    schema_version=\"1.0.18\",\n",
    "    description=sdata_zarr_artifact.description,\n",
    ")\n",
    "\n",
    "dataset_uid = \"sdata_visium\"\n",
    "dataset = vc.add_dataset(name=\"Breast Cancer Visium\", uid=dataset_uid).add_object(\n",
    "    vit.SpatialDataWrapper(\n",
    "        sdata_artifact=sdata_zarr_artifact,\n",
    "        # The following paths are relative to the root of the SpatialData zarr store on-disk.\n",
    "        image_path=\"images/CytAssist_FFPE_Human_Breast_Cancer_full_image\",\n",
    "        table_path=\"tables/table\",\n",
    "        obs_feature_matrix_path=\"tables/table/X\",\n",
    "        obs_spots_path=\"shapes/CytAssist_FFPE_Human_Breast_Cancer\",\n",
    "        region=\"CytAssist_FFPE_Human_Breast_Cancer\",\n",
    "        coordinate_system=\"global\",\n",
    "        coordination_values={\n",
    "            # The following tells Vitessce to consider each observation as a \"spot\"\n",
    "            \"obsType\": \"spot\",\n",
    "        },\n",
    "    )\n",
    ")\n",
    "\n",
    "# Add views (visualizations) to the configuration:\n",
    "spatial = vc.add_view(\"spatialBeta\", dataset=dataset)\n",
    "feature_list = vc.add_view(\"featureList\", dataset=dataset)\n",
    "layer_controller = vc.add_view(\"layerControllerBeta\", dataset=dataset)\n",
    "\n",
    "# Initialize visual properties for multiple linked views:\n",
    "vc.link_views_by_dict(\n",
    "    [spatial, layer_controller],\n",
    "    {\n",
    "        \"imageLayer\": vit.CoordinationLevel(\n",
    "            [\n",
    "                {\n",
    "                    \"photometricInterpretation\": \"RGB\",\n",
    "                }\n",
    "            ]\n",
    "        ),\n",
    "    },\n",
    "    scope_prefix=vit.get_initial_coordination_scope_prefix(dataset_uid, \"image\"),\n",
    ")\n",
    "vc.link_views([spatial, layer_controller, feature_list], [\"obsType\"], [\"spot\"])\n",
    "\n",
    "# Layout the views\n",
    "vc.layout(spatial | (feature_list / layer_controller));"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "Save the `VitessceConfig` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "sdata_vc_artifact = ln.integrations.save_vitessce_config(\n",
    "    vc,\n",
    "    description=\"View Visium SpatialData Example in Vitessce\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    ":::{note}\n",
    "\n",
    "After running `save_vitessce_config`, a Vitessce button will appear next to the dataset in the [Artifacts](https://lamin.ai/vitessce/examples/artifacts) page of the web interface.\n",
    "\n",
    "If your `VitessceConfig` object references data from multiple artifacts, the Vitessce button will appear next to a `Collection` that groups these artifacts (on the [Collections](https://lamin.ai/vitessce/examples/collections) tab of the Artifacts page).\n",
    "In the case of OME-TIFF, the presence of the corresponding offsets JSON file will result in the creation of a Collection.\n",
    "\n",
    ":::"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This creates an interactive viewer with spatial coordinates, gene expression, and tissue image:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"https://lamin-site-assets.s3.amazonaws.com/.lamindb/Xl0KgR2dZJHfmn1Q0000.png\" width=\"900px\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "## Visualize an image (OME-TIFF format)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "Vitesse can visualize data from multiple bioimaging file formats, including [OME-TIFF](https://ome-model.readthedocs.io/en/stable/ome-tiff/). \n",
    "Again, we use an example dataset that was previously ingested into the [vitessce/examples](https://lamin.ai/vitessce/examples) instance in this [transform](https://lamin.ai/vitessce/examples/transform/y4tFWARW6bJ1). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "ome_tiff_artifact = db.Artifact.get(\n",
    "    key=\"vitessce_examples/VAN0006-LK-2-85-PAS_registered.ome.tif\"\n",
    ").save()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "When using OME-TIFF files, we can use [generate-tiff-offsets](https://github.com/hms-dbmi/generate-tiff-offsets) to create an index for the bytes within the OME-TIFF file. \n",
    "We store these to a companion `.offsets.json` file which makes loading subsets of the image more efficient. \n",
    "We use a pre-generated file generated in the [transform](https://lamin.ai/vitessce/examples/transform/y4tFWARW6bJ1) above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "offsets_artifact = db.Artifact.get(\n",
    "    key=\"vitessce_examples/VAN0006-LK-2-85-PAS_registered.offsets.json\"\n",
    ").save()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "### Save a VitessceConfig object"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "You can create a dashboard for one or several datasets by using Vitessce's component API.\n",
    "\n",
    "Here, we use the [ImageOmeTiffWrapper](https://python-docs.vitessce.io/api_data.html#vitessce.wrappers.ImageOmeTiffWrapper) class to specify which pair of OME-TIFF file and offsets JSON file to load."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "vc = vit.VitessceConfig(\n",
    "    schema_version=\"1.0.18\", description=ome_tiff_artifact.description\n",
    ")\n",
    "\n",
    "dataset = vc.add_dataset(\"Image\").add_object(\n",
    "    vit.ImageOmeTiffWrapper(\n",
    "        img_artifact=ome_tiff_artifact,\n",
    "        offsets_artifact=offsets_artifact,\n",
    "    )\n",
    ")\n",
    "\n",
    "spatial = vc.add_view(\"spatialBeta\", dataset=dataset)\n",
    "layer_controller = vc.add_view(\"layerControllerBeta\", dataset=dataset)\n",
    "vc.layout(spatial | layer_controller);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "Save the `VitessceConfig` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "ome_tiff_vc_artifact = ln.integrations.save_vitessce_config(\n",
    "    vc,\n",
    "    description=\"View PAS OME-TIFF, Neumann et al., 2020 in Vitessce\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "## Visualize an image (OME-Zarr format)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "We retrieve an OME-Zarr (also known as OME-NGFF) formatted image that was ingested into [vitessce/examples](https://lamin.ai/vitessce/examples) alongside the previous datasets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "ome_zarr_artifact = db.Artifact.get(key=\"vitessce_examples/visium.ome.zarr\").save()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "### Save a VitessceConfig object"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "You can create a dashboard for one or several datasets by using Vitessce's component API.\n",
    "Here, we use the [ImageOmeZarrWrapper](https://python-docs.vitessce.io/api_data.html#vitessce.wrappers.ImageOmeZarrWrapper) class to specify an OME-Zarr file to load for visualization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "vc = vit.VitessceConfig(\n",
    "    schema_version=\"1.0.18\", description=ome_zarr_artifact.description\n",
    ")\n",
    "\n",
    "dataset_uid = \"ome_zarr_image\"\n",
    "dataset = vc.add_dataset(\"Image\", uid=dataset_uid).add_object(\n",
    "    vit.ImageOmeZarrWrapper(\n",
    "        img_artifact=ome_zarr_artifact,\n",
    "    )\n",
    ")\n",
    "\n",
    "spatial = vc.add_view(\"spatialBeta\", dataset=dataset)\n",
    "layer_controller = vc.add_view(\"layerControllerBeta\", dataset=dataset)\n",
    "vc.link_views_by_dict(\n",
    "    [spatial, layer_controller],\n",
    "    {\n",
    "        \"imageLayer\": vit.CoordinationLevel(\n",
    "            [\n",
    "                {\n",
    "                    \"photometricInterpretation\": \"RGB\",\n",
    "                }\n",
    "            ]\n",
    "        ),\n",
    "    },\n",
    "    scope_prefix=vit.get_initial_coordination_scope_prefix(dataset_uid, \"image\"),\n",
    ")\n",
    "vc.layout(spatial | layer_controller);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "Save the `VitessceConfig` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "ome_zarr_vc_artifact = ln.integrations.save_vitessce_config(\n",
    "    vc,\n",
    "    description=\"View Visium OME-Zarr Example in Vitessce\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This creates an interactive tissue image viewer:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"https://lamin-site-assets.s3.amazonaws.com/.lamindb/guM5RoXCgR65MThV0000.png\" width=\"900px\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-cell"
    ]
   },
   "outputs": [],
   "source": [
    "# compare the generated vitessce config to the public one (SpatialData)\n",
    "public_vc_json = db.Artifact.get(\"Xot2a5ZAcTW3fClG0000\").load()\n",
    "sdata_vc_json = sdata_vc_artifact.load()\n",
    "\n",
    "assert public_vc_json[\"layout\"] == sdata_vc_json[\"layout\"]\n",
    "assert public_vc_json[\"coordinationSpace\"] == sdata_vc_json[\"coordinationSpace\"]\n",
    "\n",
    "# compare the generated vitessce config json to the public one (OME-TIFF)\n",
    "public_vc_json = db.Artifact.get(\"QtF1OEtYyUe1EQ1k0000\").load()\n",
    "ome_tiff_vc_json = ome_tiff_vc_artifact.load()\n",
    "assert public_vc_json[\"layout\"] == ome_tiff_vc_json[\"layout\"]\n",
    "assert public_vc_json[\"coordinationSpace\"] == ome_tiff_vc_json[\"coordinationSpace\"]\n",
    "\n",
    "# compare the generated vitessce config to the public one (OME-Zarr)\n",
    "public_vc_json = db.Artifact.get(\"cjvX6EFrdSwsxOQl0000\").load()\n",
    "ome_zarr_vc_json = ome_zarr_vc_artifact.load()\n",
    "\n",
    "assert public_vc_json[\"layout\"] == ome_zarr_vc_json[\"layout\"]\n",
    "assert public_vc_json[\"coordinationSpace\"] == ome_zarr_vc_json[\"coordinationSpace\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [],
   "source": [
    "ln.finish()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "hide-cell"
    ]
   },
   "outputs": [],
   "source": [
    "# clean up artifacts in CI run\n",
    "ln.Collection.get(key=\"View PAS OME-TIFF, Neumann et al., 2020 in Vitessce\").delete(\n",
    "    permanent=True\n",
    ")\n",
    "\n",
    "sdata_zarr_artifact.delete(permanent=True, storage=False)\n",
    "sdata_vc_artifact.delete(permanent=True)\n",
    "\n",
    "ome_tiff_artifact.delete(permanent=True, storage=False)\n",
    "offsets_artifact.delete(permanent=True, storage=False)\n",
    "ome_tiff_vc_artifact.delete(permanent=True)\n",
    "\n",
    "ome_zarr_artifact.delete(permanent=True, storage=False)\n",
    "ome_zarr_vc_artifact.delete(permanent=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "lamindb",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
